File: android/app/src/main/kotlin/com/xmreur/prysm/TorChannelHandler.kt
package com.xmreur.prysm

import android.util.Log
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.lifecycleScope
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.launch

class TorChannelHandler(
    private val activity: FragmentActivity,
    messenger: BinaryMessenger,
    private val torController: TorController
) : MethodChannel.MethodCallHandler {

    private val channel = MethodChannel(messenger, "prysm_tor")

    fun register() {
        channel.setMethodCallHandler(this)
    }

    fun unregister() {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "startTor" -> {
                activity.lifecycleScope.launch {
                    try {
                        val bridgeConfig = parseBridgeConfig(call.arguments)
                        torController.startTor(bridgeConfig)
                        result.success(null)
                    } catch (e: Exception) {
                        Log.e("TOR", "startTor failed", e)
                        result.error("START_FAILED", e.message, null)
                    }
                }
            }

            "stopTor" -> {
                activity.lifecycleScope.launch {
                    try {
                        torController.stopTor()
                        result.success(true)
                    } catch (e: Exception) {
                        Log.e("TOR", "stopTor failed", e)
                        result.error("STOP_FAILED", e.message, null)
                    }
                }
            }

            "getCachedOnionAddress" -> {
                result.success(torController.getCachedOnionAddress())
            }

            "getOnionAddress" -> {
                torController.getOnionAddressAsync { onionAddress ->
                    if (onionAddress != null && onionAddress.endsWith(".onion")) {
                        result.success(onionAddress)
                    } else {
                        result.error("NO_ADDRESS", "ONION address not available", null)
                    }
                }
            }

            else -> result.notImplemented()
        }
    }

    private fun parseBridgeConfig(arguments: Any?): TorBridgeConfig {
        val map = arguments as? Map<*, *> ?: return TorBridgeConfig(false, emptyList())
        val useObfs4 = map["useObfs4"] as? Boolean ?: false
        @Suppress("UNCHECKED_CAST")
        val bridges = map["bridges"] as? List<String> ?: emptyList()
        return TorBridgeConfig(useObfs4, bridges)
    }
}

File: ios/Runner/Tor/TorChannelHandler.swift
import Flutter
import Foundation

final class TorChannelHandler {
    private static let channelName = "prysm_tor"

    private let torController = PrysmTorController()
    private var channel: FlutterMethodChannel?

    static func register(messenger: FlutterBinaryMessenger) {
        let handler = TorChannelHandler()
        let channel = FlutterMethodChannel(
            name: channelName,
            binaryMessenger: messenger
        )
        handler.channel = channel
        channel.setMethodCallHandler(handler.handle)
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startTor":
            Task {
                do {
                    let bridgeConfig = Self.parseBridgeConfig(call.arguments)
                    try await torController.startTor(bridgeConfig: bridgeConfig)
                    result(nil)
                } catch {
                    NSLog("TOR startTor failed: \(error)")
                    result(
                        FlutterError(
                            code: "START_FAILED",
                            message: error.localizedDescription,
                            details: nil
                        )
                    )
                }
            }

        case "stopTor":
            Task {
                do {
                    try await torController.stopTor()
                    result(true)
                } catch {
                    NSLog("TOR stopTor failed: \(error)")
                    result(
                        FlutterError(
                            code: "STOP_FAILED",
                            message: error.localizedDescription,
                            details: nil
                        )
                    )
                }
            }

        case "restartTor":
            Task {
                do {
                    try await torController.restartTor()
                    result(nil)
                } catch {
                    NSLog("TOR restartTor failed: \(error)")
                    result(
                        FlutterError(
                            code: "RESTART_FAILED",
                            message: error.localizedDescription,
                            details: nil
                        )
                    )
                }
            }

        case "getCachedOnionAddress":
            Task {
                result(await torController.getCachedOnionAddress())
            }

        case "getOnionAddress":
            Task {
                if let address = await torController.getOnionAddress(),
                   address.hasSuffix(".onion") {
                    result(address)
                } else {
                    result(
                        FlutterError(
                            code: "NO_ADDRESS",
                            message: "ONION address not available",
                            details: nil
                        )
                    )
                }
            }

        case "setCallAudioActive":
            let active = call.arguments as? Bool ?? false
            TorKeepAlive.shared.setCallAudioActive(active)
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private static func parseBridgeConfig(_ arguments: Any?) -> TorBridgeConfig {
        guard let map = arguments as? [String: Any] else {
            return TorBridgeConfig(useObfs4: false, bridges: [])
        }
        let useObfs4 = map["useObfs4"] as? Bool ?? false
        let bridges = map["bridges"] as? [String] ?? []
        return TorBridgeConfig(useObfs4: useObfs4, bridges: bridges)
    }
}
